### DISCUSSION TITLE
Allow overriding `icon` for entities which set a `device_class`

### DISCUSSION THREAD

**Discussion by emontnemery** [home-assistant/architecture#552] [Entity Models] [answered]
According to the developer documentation for `Entity.icon`:

> Icon to use in the frontend. Icons start with mdi: plus an identifier. Icon and device_class are mutually exclusive; Do not set an icon if a device_class is present. You probably don't need this since Home Assistant already provides default icons for all devices.

Proposal:
Reword the documentation for `Entity.icon` to make it allowed, but not encouraged, to override the default icon. An example could be a heating system integration exposing the setpoints for day and night temperatures, where icons conveying day and night may be better than the default thermometer icon for `DEVICE_CLASS_TEMPERATURE`.

Alternative:
Extend the device classes to be more precise. Using `DEVICE_CLASS_TEMPERATURE` as example, there are many cases where a sensor may expose a temperature, and a single device class for all those cases is very broad.

**Comment by balloob** [comment]
There are two reasons why the icon won't match:

 - The device class is too broad and it covers multiple things (weather temperature, CPU temperature, pool temperature)
 - The device has properties that don't match the standard variant, and so the icon is off (ie a vertical vs horizontal opening window)

**Comment by MartinHjelmare** [comment]
I think it's ok to allow custom icons for the rare cases when there's no better category possible. But I'd like us to not allow this to become a simple way out to avoid the very important task and problem that Home Assistant is solving, which is standardization of devices and services.

A complimentary solution that I'd like to see is to add a label per device or service. The existing device class on entity level and the new device/service label would together be able to better describe the measurement feature.

One example where this could be an improvement is in the frontend where depending on the context the frontend could decide to show either the entity device class icon or the new device/service category icon.

Eg when looking at the device page it could be more interesting to distinguing the entity device classes with corresponding icons than showing the same device specific icon for all features. While on a panel with a mix of different devices it could be better to show the device specific icon.

Weather service pressure measurements could be resolved from vessel device pressure measurements.

Weather sensor measurements could automatically be grouped in a frontend card without needing a weather platform.

I think there are many opportunities to be had from adding such a label.

**Reply by thomasloven** [reply]
I'm a strong proponent of labels/tags. Another use case would be to label a measurement as "technical" or something, so e.g. people who normally use Fahrenheit can still have their 3d printer nozzle temperature displayed in Celsius.

**Reply by MartinHjelmare** [reply]
Alternatively we could limit the existing `device_class` property to only return actual device/service labels and move the entity category, eg pressure, to a new property. Currently we're mixing different kinds of categorization into `device_class` which is not good. Eg device class `tv` identifies the device but device class `temperature` doesn't identify a device.

**Reply by emontnemery** [reply]
> we could limit the existing device_class property to only return actual device/service labels and move the entity category, eg pressure, to a new property

I think this sounds very good. Consider a heating system for example, which may expose multiple temperature sensors. It then does not help all that much to know that the device is a heating system, we would also need to know something about the individual temperature sensors.

**Comment by emontnemery** [comment, marked as answer]
PR to update the wording for `Entity.icon` since we seem to agree we can tolerate overriding it: home-assistant/developers.home-assistant#894

**Reply by balloob** [reply]
Approved. I think that we don't want to be forced to add a category for every single type of data in the world with a corresponding icon. This will bloat our data spec and causes unnecessary noise.
